pennylane_forest/converter.py

```diff
@@ -0,0 +1,626 @@
+import pyquil
+import pyquil.gates as g
+
+pyquil_inv_operation_map = {
+    "I": qml.Identity,
+    "X": qml.PauliX,
+    "Y": qml.PauliY,
+    "Z": qml.PauliZ,
+    "H": qml.Hadamard,
+    "CNOT": qml.CNOT,
+    "SWAP": qml.SWAP,
+    "CZ": qml.CZ,
+    "PHASE": qml.PhaseShift,
+    "RX": qml.RX,
+    "RY": qml.RY,
+    "RZ": qml.RZ,
+    "CRX": qml.CRX,
+    "CRY": qml.CRY,
+    "CRZ": qml.CRZ,
+    "S": plf.ops.S,
+    "T": plf.ops.T,
+    "CCNOT": plf.ops.CCNOT,
+    "CPHASE": lambda *params, wires: plf.ops.CPHASE(*params, 3, wires=wires),
+    "CPHASE00": lambda *params, wires: plf.ops.CPHASE(*params, 0, wires=wires),
+    "CPHASE01": lambda *params, wires: plf.ops.CPHASE(*params, 1, wires=wires),
+    "CPHASE10": lambda *params, wires: plf.ops.CPHASE(*params, 2, wires=wires),
+    "CSWAP": plf.ops.CSWAP,
+    "ISWAP": plf.ops.ISWAP,
+    "PSWAP": plf.ops.PSWAP,
+}
+
+_control_map = {
+    "X": "CNOT",
+    "Z": "CZ",
+    "CNOT": "CCNOT",
+    "SWAP": "CSWAP",
+    "RX": "CRX",
+    "RY": "CRY",
+    "RZ": "CRZ",
+    "PHASE": "CPHASE",
+}
+
+_matrix_dictionary = pyquil.gate_matrices.QUANTUM_GATES
+
+
+def _direct_sum(A, B):
+    """Return the direct sums of two arrays.
+    
+    Args:
+        A (np.array): The first array (upper left array)
+        B (np.array): The second array (lower right array)
+    
+    Returns:
+        np.array: The direct sum of the two input arrays
+    """
+    sum = np.zeros(np.add(A.shape, B.shape), dtype=A.dtype)
+    sum[: A.shape[0], : A.shape[1]] = A
+    sum[A.shape[0] :, A.shape[1] :] = B
+
+    return sum
+
+
+def _controlled_matrix(op):
+    """Return the matrix associated with the controlled operation.
+    
+    Args:
+        op (np.array): Array representing the operation 
+                       that should be controlled
+    
+    Returns:
+        np.array: Array representing the controlled operations. If the input
+                  array has shape (N, N) the output shape is (2*N, 2*N).
+    """
+    return _direct_sum(np.eye(op.shape[0], dtype=op.dtype), op)
+
+
+def _resolve_gate(gate):
+    """Resolve the given pyquil Gate as far as possible.
+
+    For example, the gate `CONTROLLED CONTROLLED X` will be resolved to `CCNOT`. 
+    The gate `CONTROLLED CONTROLLED RX(0.3)` will be resolved to `CONTROLLED CRX(0.3)`.
+    
+    Args:
+        gate (pyquil.quil.Gate): The gate that should be resolved
+    
+    Returns:
+        pyquil.quil.Gate: The maximally resolved gate
+    """
+    for i, modifier in enumerate(gate.modifiers):
+        if modifier == "CONTROLLED":
+            if gate.name in _control_map:
+                stripped_gate = g.Gate(_control_map[gate.name], gate.params, gate.qubits)
+                stripped_gate.modifiers = gate.modifiers.copy()
+                del stripped_gate.modifiers[i]
+
+                return _resolve_gate(stripped_gate)
+            else:
+                break
+
+    return gate
+
+
+def _resolve_params(params, variable_map):
```

Comment:
If I understand correctly, the `variable_map` would correspond to a `memory_map` used when calling ``qc.run()``  (example can be found at [section 1.4.9 Using memory_map](https://buildmedia.readthedocs.org/media/pdf/pyquil/latest/pyquil.pdf))

Is the distinction between these two terms in naming intentional? If so, what would the reason be for it?

Reply:
I actually didn't know about `memory_map`. But as this is pennylane-facing code I would not use rigetti naming for this. 

I'll rename it to `parameter_map` because I think parameter is what we mostly coll them in PennyLane.

Reply:
I didn't go for `params` intentionally, because this somewhat obfuscates the fact that I require a dictionary.